grpclb/src/main/java/io/grpc/grpclb/GrpclbState.java

```diff
@@ -772,48 +772,52 @@ private void cleanUp() {
   private void maybeUpdatePicker() {
     List<RoundRobinEntry> pickList;
     ConnectivityState state;
-    switch (config.getMode()) {
-      case ROUND_ROBIN:
-        pickList = new ArrayList<>(backendList.size());
-        Status error = null;
-        boolean hasPending = false;
-        for (BackendEntry entry : backendList) {
-          Subchannel subchannel = entry.subchannel;
-          Attributes attrs = subchannel.getAttributes();
-          ConnectivityStateInfo stateInfo = attrs.get(STATE_INFO).get();
-          if (stateInfo.getState() == READY) {
-            pickList.add(entry);
-          } else if (stateInfo.getState() == TRANSIENT_FAILURE) {
-            error = stateInfo.getStatus();
-          } else {
-            hasPending = true;
-          }
-        }
-        if (pickList.isEmpty()) {
-          if (error != null && !hasPending) {
-            pickList.add(new ErrorEntry(error));
-            state = TRANSIENT_FAILURE;
-          } else {
-            pickList.add(BUFFER_ENTRY);
-            state = CONNECTING;
+    if (backendList.isEmpty()) {
+      if (balancerWorking)  {
+        pickList =
+            Collections.<RoundRobinEntry>singletonList(
+                new ErrorEntry(NO_AVAILABLE_BACKENDS_STATUS));
+        state = TRANSIENT_FAILURE;
+      } else if (usingFallbackBackends) {
+        pickList =
+            Collections.<RoundRobinEntry>singletonList(
+                new ErrorEntry(NO_FALLBACK_BACKENDS_FOUND_STATUS));
+        state = TRANSIENT_FAILURE;
+      } else {  // still waiting for balancer
+        pickList = Collections.singletonList(BUFFER_ENTRY);
+        state = CONNECTING;
+      }
```

Comment:
You could just call maybeUpdatePicker and return right after this to avoid putting all the round-robin/pick-first logic in the else? That would greatly reduce amount of indentation and make it easier to read.

Reply:
Fix.